src/erldns_handler.erl

```diff
@@ -48,24 +50,37 @@ start_link() ->
 %% @doc Register a record handler.
 -spec register_handler([dns:type()], module()) -> ok.
 register_handler(RecordTypes, Module) ->
-  gen_server:call(?MODULE, {register_handler, RecordTypes, Module}).
+  register_handler(RecordTypes, Module, ?DEFAULT_HANDLER_VERSION).
+
+%% @doc Register a record handler with version.
+-spec register_handler([dns:type()], module(), integer()) -> ok.
+register_handler(RecordTypes, Module, Version) ->
+  gen_server:call(?MODULE, {register_handler, RecordTypes, Module, Version}).
 
 %% @doc Get all registered handlers along with the DNS types they handle.
 -spec get_handlers() -> [{module(), [dns:type()]}].
 get_handlers() ->
-  gen_server:call(?MODULE, {get_handlers}).
+  get_handlers(?DEFAULT_HANDLER_VERSION).
 
+%% @doc Get all registered handlers along with the DNS types they handle + version
+-spec get_handlers(integer()) -> [{module(), [dns:type()]}].
+get_handlers(Version) ->
+  gen_server:call(?MODULE, {get_handlers, Version}).
 
 % gen_server callbacks
 
 init([]) ->
   {ok, #state{handlers=[]}}.
 
 handle_call({register_handler, RecordTypes, Module}, _, State) ->
-  %lager:info("Registered handler (module: ~p, types: ~p)", [Module, RecordTypes]),
-  {reply, ok, State#state{handlers = State#state.handlers ++ [{Module, RecordTypes}]}};
-handle_call({get_handlers}, _, State) ->
-  {reply, State#state.handlers, State}.
+  lager:info("Registered handler (module: ~p, types: ~p)", [Module, RecordTypes]),
+  {reply, ok, State#state{handlers = State#state.handlers ++ [{Module, RecordTypes, 1}]}};
+handle_call({register_handler, RecordTypes, Module, Version}, _, State) ->
+  lager:info("Registered handler (module: ~p, types: ~p, version: ~p)", [Module, RecordTypes, Version]),
+  {reply, ok, State#state{handlers = State#state.handlers ++ [{Module, RecordTypes, Version}]}};
+handle_call({get_handlers, Version}, _, State) ->
+  VersionHandlers = [erlang:delete_element(3, X) || X <- State#state.handlers, element(3, X) =:= Version],
```

Comment:
Rather than stripping out version information, does it make sense to return it and ensure all callers can properly handle the new data structure?

Reply:
done intentionally to preserve backwards compatibility.

my initial implementation was carrying over version information, but the pattern matching + function returns needs to be looked after and custom_lookup was failing with bad args.

get_handlers/0 and /1 always will return the proper list of handlers without changing additional code - pure separation of concerns and minimum impact on existing code was the target.

If this changes - first point to change is:

https://github.com/dnsimple/erldns/blob/dev_ecs/src/erldns_resolver.erl#L486

But maybe there are more to track and refactor.

Reply:
I think there will need to be a way to get handlers with the version number in the response in order to support logic branching in `erldns_resolver` based on the version number.

If backwards compatibility is absolutely necessary, then strip the version in the `get_handlers` implementation, not in the `gen_server` callbacks. That way you can add a new function `get_versioned_handlers` (or something like that) and include the version in the return value.

Reply:
commit ad80dba04c8c6b290580fccacc796f021e3aeaa2 (HEAD -> dev_ecs, origin/dev_ecs)
Author: atanas argirov <atanas@argirov.org>
Date:   Wed Mar 11 14:32:25 2020 +0000

    * added get_versioned_handlers/0
    * get_handlers/0 returns all handlers with version set to 1
    * new handlers will need to use get_versioned_handlers/0 function

the commit above addresses handlers's part.

A/AAAA qtypes working fine, need to address ANYs and complete the rest of erldnsimple plugins